jobs/base_model_validator.py

```diff
@@ -0,0 +1,171 @@
+# coding: utf-8
+#
+# Copyright 2021 The Oppia Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS-IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Beam functions and transforms to provide validation for models. The
+BaseModelValidator is intended to be a class that other validators can inherit
+from. It takes in a Beam PCollection of models and returns a PCollection of the
+validation errors found in the input. The Beam.DoFn classes are functions that
+are called in the BaseModelValidator to perform validations.
+
+When writing subclasses to BaseModelValidator, call the new added
+validation functions in the expand function, and then flatten the output
+with the result of the super function.
+"""
+
+from __future__ import absolute_import  # pylint: disable=import-only-modules
+from __future__ import unicode_literals  # pylint: disable=import-only-modules
+
+import datetime
+import re
+
+from core.domain import cron_services
+from core.platform import models
+from jobs import base_model_validator_errors as errors
+from jobs import jobs_utils
+
+import apache_beam as beam
+
+(base_models, user_models) = models.Registry.import_models(
+    [models.NAMES.base_model, models.NAMES.user])
+
+
+MAX_CLOCK_SKEW_SECS = datetime.timedelta(seconds=1)
+
+
+class ValidateModelIdWithRegex(beam.DoFn):
+    """DoFn to validate model ids against a given regex string."""
+
+    def __init__(self, regex_string):
```

Comment:
I've learned that we can't depend on constructors in Apache Beam, it breaks the user code contract ([source](https://beam.apache.org/documentation/programming-guide/#requirements-for-writing-user-code-for-beam-transforms)).

The fix is to use side inputs instead, which is pretty straight foward:

```python
class ValidateModelIdWithRegex(beam.DoFn):
    def process(self, input_model, regex_string=None):
        if not re.match(regex_string, model.id):
            yield errors.ModelInvalidIdError(model)
```

And we can pass it in the call site as an extra kwarg of `ParDo`:

```python
class BaseModelValidator(beam.PTransform):
    def expand(self, model_pipe):
        model_id_validation_errors = (
            model_pipe
            | beam.ParDo(
                ValidateModelIdWithRegex(),
                regex_string=self._get_model_id_regex()))

Reply:
Done